solidity/contracts/ECDSARewardsDistributor.sol

```diff
@@ -0,0 +1,134 @@
+/**
+▓▓▌ ▓▓ ▐▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▄
+▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+  ▓▓▓▓▓▓    ▓▓▓▓▓▓▓▀    ▐▓▓▓▓▓▓    ▐▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+  ▓▓▓▓▓▓▄▄▓▓▓▓▓▓▓▀      ▐▓▓▓▓▓▓▄▄▄▄         ▓▓▓▓▓▓▄▄▄▄         ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+  ▓▓▓▓▓▓▓▓▓▓▓▓▓▀        ▐▓▓▓▓▓▓▓▓▓▓▌        ▓▓▓▓▓▓▓▓▓▓▌        ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+  ▓▓▓▓▓▓▀▀▓▓▓▓▓▓▄       ▐▓▓▓▓▓▓▀▀▀▀         ▓▓▓▓▓▓▀▀▀▀         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▀
+  ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+
+                           Trust math, not hardware.
+*/
+
+pragma solidity 0.5.17;
+
+import "@keep-network/keep-core/contracts/utils/BytesLib.sol";
+import "@keep-network/keep-core/contracts/KeepToken.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/SafeERC20.sol";
+import "openzeppelin-solidity/contracts/cryptography/MerkleProof.sol";
+
+/// @title ECDSA Rewards distributor
+/// @notice This contract can be used by stakers to claim their rewards for
+/// participation in the keep network for operating ECDSA nodes.
+/// @dev This contract is based on the Uniswap's Merkle Distributor
+/// https://github.com/Uniswap/merkle-distributor with some modifications:
+/// - added a map of merkle root keys with the amount of KEEP (value) that will
+///   be allocated for those merkle roots
+/// - added receiveApproval() function that will be called each time to allocate
+///   new KEEP rewards for a given merkle root. Merkle root is going to be generated
+///   regulary (ex. every week) and it is also means that an interval for that
+///   merkle root has passed
+/// - changed code accordingly to process claimed rewards using a map of merkle
+///   roots
+contract ECDSARewardsDistributor is Ownable {
+    using SafeERC20 for KeepToken;
+    using BytesLib for bytes;
+    using SafeMath for uint256;
+
+    KeepToken public token;
+
+    // This event is triggered whenever a call to #claim succeeds.
+    event RewardsClaimed(uint256 index, address account, uint256 amount);
+    // This event is triggered whenever rewards are allocated.
+    event RewardsAllocated(bytes32 merkleRoot, uint256 amount);
+
+    // Merkle root -> total amount for distribution for a given timeframe.
+    mapping(bytes32 => uint256) private merkleRoots;
```

Comment:
There is a delicate naming problem with this field, how it is modified in `claim` function, and how it is exposed to the world by `getAllocation`.

If I call `getAllocation`, I expect to receive an original reward allocation for the given merkle root (interval). This is not what I am getting because of `sub` in `claim` function. So one option could be to remove `sub` instruction in `claim`. On the other hand, given that this value is written once, it never changes and does not provide any valuable information to the contract, I think we can get rid of `getAllocation` as it can be read now from `RewardsAllocated` event, update `mapping(bytes32 => uint256) private merkleRoots;` to `mapping(bytes32 => bool) private merkleRoots;`, and update the `require` in `claim` accordingly. What we are really interested in is whether the given `merkleRoot` was blessed by the contract owner, that's all. Also, this change will make weekly allocations a bit cheaper.

Reply:
> I think we can get rid of getAllocation as it can be read now from RewardsAllocated event

but if for some reason you would like to check remaining KEEP for a given `merkleRoot` then how would you do this? I agree regarding `RewardsAllocated` event that carry data on the initial funding for a 'merkleRoot' (interval), but having an additional function that would check remaining KEEP after claiming some rewards might be handy. We can rename it ex. `getRemainingAllocation(bytes32 merkleRoot)`. I might be missing some product rules here, so this is just to double check that we don't need this function.

Reply:
From the staker perspective, you need to inspect events anyway to see if you have already claimed rewards for the given interval and it's probably just one call, assuming we index event fields properly.

I do agree, seeing the number of tokens that have not been claimed yet for the given interval without having to inspect events could be handy *for us* in some scenarios but... in the same scenarios, we would have to inspect events anyway to see which stakers haven't claimed their rewards yet so that we can maybe ping them.

Having `getRemainingAllocation` - or maybe even better `unclaimedRewards` given that we have `claim` and `isClaimed` - could help us start the inspection. On the other hand:
- For small stakers, gas costs can be crucial. We have already experienced a situation with extremely high gas prices and the cost of beacon rewards withdrawal to be the same as the reward for some small stakers.
- If we do not count not-yet-claimed rewards here, we do not have to worry whether we should allow adding tokens for a given merkle root ([your comment](https://github.com/keep-network/keep-ecdsa/pull/627#discussion_r534035982)). If we do a mistake, we'll just send more tokens. The issue we are having is that we are duplicating the same information in two places - amounts are already represented in the merkle root and claimed bit map and at the same time, we have a separate counter that is also tracking them.

Reply:
Okay, agree, bottom line is that we can remove tracking of `merkleRoot -> (unclaimed)amount`. 04bf914